experimental/packages/opentelemetry-sdk-metrics-base/src/export/MetricReader.ts

```diff
@@ -15,51 +15,206 @@
  */
 
 import { AggregationTemporality } from './AggregationTemporality';
-import { MetricExporter } from './MetricExporter';
 import { MetricProducer } from './MetricProducer';
+import { MetricData } from './MetricData';
+import { ReaderResult, ReaderResultCode } from './ReaderResult';
+
+export type ReaderOptions<T> = {
+  done?: (result: ReaderResult<T>) => void
+  timeoutMillis?: number
+}
+
+export type ReaderCollectionOptions = ReaderOptions<MetricData[]>;
+
+export type ReaderShutdownOptions = ReaderOptions<void>;
+
+export type ReaderForceFlushOptions = ReaderOptions<void>;
+
+
+/**
+ * Adds a timeout to a promise and executes the callback if the specified timeout has elapsed, or the promise
+ * has resolved or rejected.
+ *
+ * <p> NOTE: this operation will continue even after the timeout fires the callback.
+ *
+ * @param promise promise to use with timeout.
+ * @param timeout the timeout in milliseconds until the returned promise is rejected.
+ * @param done the callback once the promise has resolved or rejected.
+ */
+export function promiseWithTimeout<T>(promise: Promise<T>, timeout: number, done: (result: ReaderResult<T>) => void): void {
```

Comment:
Maybe we can move this function to the src/utils.ts so that we can facilitate it in observable callbacks too?

Reply:
I changed this since the code is now promise-based. But it might still be useful in utils. Should I still move it? :thinking:

Reply:
+1 for moving it. If so, the ReaderTimeoutError should also be renamed as a generic TimeoutError if possible.

Reply:
Moved this and renamed `ReaderTimeoutError` to `TimeoutError` in 1ec3150, thanks for the feedback!